net: l2: ieee802154: L2/L3 decoupling

fgrandel opened the pull request:
This fixes #48585.

It also contains some minor documentation and formatting improvements - see individual commits.

Commit:
net: l2: ieee802154: Improve TODO marker consistency

Most existing TODO markers (as well as VSCode default settings) favor
TODO over ToDo - so let's make this a little bit more consistent in the
IEEE 802.15.4 module.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: improve inline comments

This change contains some merely editorial changes to inline comments
plus updates references from the IEEE 802.15.4-2003 spec to
IEEE 802.15.4-2006 which corresponds to the implementation level of
the module.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: decouple L2/L3 concerns

This change decouples the IEEE 802.15.4 (L2) layer from all IPv6 (L3)
concerns.

Applications may now choose to set CONFIG_NET_6LO=n and
CONFIG_NET_L2_IEEE802154=y at the same time.

Setting CONFIG_NET_6LO=n will build a vanilla IEEE 802.15.4-2006 specs
compliant L2 layer without any reference to 6LoWPAN or IPv6. This allows
application developers to design custom non-IP protocols on top of
IEEE 802.15.4-2006 and thereby makes the L2 layer much more re-usable.

Fixes #48585.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: rename *fragment to *6lo_fragment

6LoWPAN fragmentation is not related to IEEE 802.15.4 proper but is just
part of its IPv6-specific L3-adaptation layer. To make this more obvious
we rename all resources related to 6LoWPAN fragmentation.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

Commit:
net: l2: ieee802154: apply clang format

This change makes the files which are part of this changeset comply to
the project's coding style rules as defined in .clang-format.

This required addition of some forward declarations and additional
dependencies into header files as some of them depended on the order of
header inclusion which was changed due to alphabetical ordering of
includes.

Background: .clang-format states "SortIncludes:true" which will force
re-ording of include-statements which in turn might break the build if
header file inclusion is not order-independent.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

nandojve reviewed (changes requested):
Nice initiative @fgrandel !

fgrandel commented on subsys/net/l2/ieee802154/ieee802154_fragment.c:
```diff
@@ -1,5 +1,5 @@
 /** @file
- * @brief 802.15.4 fragment related functions
+ * @brief 802.15.4 6LoWPAN fragment related functions
```
Fixed.

rlubos commented on .clang-format:
```diff
@@ -77,6 +77,7 @@ IndentCaseLabels: false
 IndentWidth: 8
 # SpaceBeforeParens: ControlStatementsExceptControlMacros # clang-format >= 13.0
 UseTab: Always
+SortIncludes: false
```
Shouldn't the header dependencies be fixed instead? It doesn't seem right that headers need to be included in certain order to work.

fgrandel replied:
Agreed - will work on it.

rlubos reviewed (commented):
Looks pretty neat overall, just one concern.

gmarull commented on .clang-format:
```diff
@@ -77,6 +78,7 @@ IndentCaseLabels: false
 IndentWidth: 8
 # SpaceBeforeParens: ControlStatementsExceptControlMacros # clang-format >= 13.0
 UseTab: Always
+SortIncludes: false
```
-1, fix headers instead please.

fgrandel commented on subsys/net/l2/ieee802154/ieee802154.c:
```diff
@@ -166,8 +161,8 @@ static bool ieeee802154_check_dst_addr(struct net_if *iface,
-		if (memcmp(dst_plain->addr.ext_addr, ctx->ext_addr,
-					IEEE802154_EXT_ADDR_LENGTH) != 0) {
+		if (memcmp(dst_plain->addr.ext_addr, ctx->ext_addr, IEEE802154_EXT_ADDR_LENGTH) !=
+		    0) {
```
While I totally agree this has larger repercussions - I therefore opened an RFC for better visibility of the question... see #48616

fgrandel replied:
@jukkar, is it ok, to leave this as-is in this PR and prepare a separate PR for the general formatting change as proposed by @gmarull in #48616? This would decouple the more general discussion from this changeset.

fgrandel replied:
Fixed manually for the moment being (breaks automatic formatting) as this shouldn't hold up the PR. Remainder to be discussed in #48616 and fixed in a separate PR.

fgrandel commented:
Unfortunately some commits had an "author" field with an email that no longer exists. Fixed the author fields now (this is the only change of the new push).

rlubos reviewed (dismissed):
LGTM, thanks!

Commit:
net: ip: only start IP configuration on IP interfaces

net_if_up() unconditionally calls iface_ipv6_start() and
net_ipv4_autoconf_start(). This is not correct for interfaces that do
not support IPv4/v6.

This patch therefore introduces a check on the interface's flags before
proceeding.

Signed-off-by: Florian Grandel <jerico.dev@gmail.com>

fgrandel commented:
While testing this stuff on a real device I found more places where the assumption was made that all interfaces necessarily have an IP configured. Could you please review those changes again?

fgrandel commented:
Thanks everyone, this was a great first contributor experience. Makes me want to contribute much more to the Zephyr project. My holidays are almost over so now its the weekend schedule.
